### PR TITLE
Add ValidationRuleSet.Remove

### PR DESCRIPTION
**Pull request by LucGenetier**

### CODE REVIEW

File: src/Microsoft.OpenApi/Validations/ValidationRule.cs [hunk 1]
```diff
@@ -36,13 +40,29 @@ public class ValidationRule<T> : ValidationRule where T : IOpenApiElement
...
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ValidationRule"/> class.
+        /// </summary>
+        /// <param name="name">Validation rule name.</param>
+        /// <param name="validate">Action to perform the validation.</param>
+        public ValidationRule(string name, Action<IValidationContext, T> validate)
         {
             _validate = Utils.CheckArgumentNull(validate);
+            Name = name ?? "UnnamedRule";
```
Comment (baywet) [inline src/Microsoft.OpenApi/Validations/ValidationRule.cs:58]: use string.IsNullOrEmpty with ternary operator instead.

**Comment by baywet** [comment]
Can you also rebase onto vnext please so we don't get all this merge commits in the way?

File: src/Microsoft.OpenApi/Validations/ValidationRuleSet.cs [hunk 2]
```diff
@@ -134,6 +134,28 @@ public void Add(ValidationRule rule)
             item.Add(rule);
         }
 
+        /// <summary>
+        /// Remove a rule by name.
+        /// </summary>        
+        /// <param name="ruleName">Name of the rule.</param>
+        public void Remove(string ruleName)
```
Comment (baywet) [inline src/Microsoft.OpenApi/Validations/ValidationRuleSet.cs:141]: Having an overload that allows people to provide element type as well might be useful

File: src/Microsoft.OpenApi/Validations/ValidationRule.cs [hunk 3]
```diff
@@ -20,14 +20,19 @@ public abstract class ValidationRule
...
         /// <summary>
         /// Validate the object.
         /// </summary>
         /// <param name="context">The context.</param>
         /// <param name="item">The object item.</param>
         internal abstract void Evaluate(IValidationContext context, object item);
+
+        internal ValidationRule(string name)
+        {
+            Name = name;
```
Comment (baywet) [inline src/Microsoft.OpenApi/Validations/ValidationRule.cs:34]: please check for null or empty here

File: src/Microsoft.OpenApi/Validations/ValidationRule.cs [hunk 4]
```diff
@@ -37,11 +42,12 @@ public abstract class ValidationRule
...
         private readonly Action<IValidationContext, T> _validate;
-
+        
         /// <summary>
         /// Initializes a new instance of the <see cref="ValidationRule"/> class.
         /// </summary>        
         /// <param name="validate">Action to perform the validation.</param>
+        [Obsolete("Please use the other constructor and specify a name")]
         public ValidationRule(Action<IValidationContext, T> validate)
             : this (null, validate)
```
Comment (baywet) [inline src/Microsoft.OpenApi/Validations/ValidationRule.cs:52]:
```suggestion
            : this (Guid.NewGuid().ToString("D"), validate)
```

This way we avoid any side-effect during the remove call.

**Review by baywet — CHANGES_REQUESTED** [review]
Thanks for making the changes, a couple more recommendations

File: src/Microsoft.OpenApi/Validations/ValidationRule.cs [hunk 5]
```diff
@@ -18,6 +17,11 @@ public abstract class ValidationRule
         /// </summary>
         internal abstract Type ElementType { get; }
 
+        /// <summary>
+        /// Validation rule Name.
+        /// </summary>
+        public abstract string Name { get; }
```
[the remark this answers is not in the document]
Comment (LucGenetier) [reply src/Microsoft.OpenApi/Validations/ValidationRule.cs]: nope

**Comment by LucGenetier** [comment]
ok

---
In reply to: [1904236027](https://github.com/microsoft/OpenAPI.NET/pull/1535#issuecomment-1904236027) [](http://example.com/codeflow?ancestors=1904236027)

File: src/Microsoft.OpenApi/Validations/ValidationRule.cs [hunk 6]
```diff
@@ -36,13 +40,29 @@
...
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ValidationRule"/> class.
+        /// </summary>
+        /// <param name="name">Validation rule name.</param>
+        /// <param name="validate">Action to perform the validation.</param>
+        public ValidationRule(string name, Action<IValidationContext, T> validate)
         {
             _validate = Utils.CheckArgumentNull(validate);
+            Name = name ?? "UnnamedRule";
```
[the remark this answers is not in the document]
Comment (LucGenetier) [reply src/Microsoft.OpenApi/Validations/ValidationRule.cs]: ok

File: src/Microsoft.OpenApi/Validations/ValidationRule.cs [hunk 7]
```diff
@@ -36,13 +40,29 @@ public class ValidationRule<T> : ValidationRule where T : IOpenApiElement
 
         /// <summary>
         /// Initializes a new instance of the <see cref="ValidationRule"/> class.
-        /// </summary>
+        /// </summary>        
         /// <param name="validate">Action to perform the validation.</param>
         public ValidationRule(Action<IValidationContext, T> validate)
```
[the remark this answers is not in the document]
Comment (LucGenetier) [reply src/Microsoft.OpenApi/Validations/ValidationRule.cs:51]: done

File: src/Microsoft.OpenApi/Validations/ValidationRule.cs (hunk 1 above)
Comment (LucGenetier) [reply src/Microsoft.OpenApi/Validations/ValidationRule.cs]: ok

File: src/Microsoft.OpenApi/Validations/ValidationRuleSet.cs [hunk 8]
```diff
@@ -134,6 +134,28 @@ public void Add(ValidationRule rule)
...
+        {
+            foreach (KeyValuePair<Type, IList<ValidationRule>> rule in _rules)
+            {
+                var validationRule = rule.Value.FirstOrDefault(vr => vr.Name.Equals(ruleName, StringComparison.Ordinal));
```
[the remark this answers is not in the document]
Comment (LucGenetier) [reply src/Microsoft.OpenApi/Validations/ValidationRuleSet.cs]: I'll fix that in deleting all rules w/ the "ruleName", unique or not

File: src/Microsoft.OpenApi/Validations/ValidationRuleSet.cs (hunk 2 above)
Comment (LucGenetier) [reply src/Microsoft.OpenApi/Validations/ValidationRuleSet.cs:141]: added

File: src/Microsoft.OpenApi/Validations/ValidationRule.cs (hunk 3 above)
Comment (LucGenetier) [reply src/Microsoft.OpenApi/Validations/ValidationRule.cs:34]: done

File: src/Microsoft.OpenApi/Validations/ValidationRule.cs (hunk 4 above)
Comment (LucGenetier) [reply src/Microsoft.OpenApi/Validations/ValidationRule.cs:52]: will do that

File: src/Microsoft.OpenApi/Validations/ValidationRuleSet.cs [hunk 9]
```diff
@@ -134,6 +134,30 @@ public void Add(ValidationRule rule)
...
+            {
+                _rules[rule.Key] = rule.Value.Where(vr => !vr.Name.Equals(ruleName, StringComparison.Ordinal)).ToList();                
+            }
+
+            // Remove types with no rule
+            _rules = _rules.Where(r => r.Value.Any()).ToDictionary(r => r.Key, r => r.Value);
+        }
+
+        /// <summary>
+        /// Remove a rule by type.
```
Comment (baywet) [inline src/Microsoft.OpenApi/Validations/ValidationRuleSet.cs:153]:
```suggestion
        /// Remove a rule by element type.
```

File: src/Microsoft.OpenApi/Validations/ValidationRuleSet.cs [hunk 10]
```diff
@@ -134,6 +134,30 @@ public void Add(ValidationRule rule)
             item.Add(rule);
         }
 
+        /// <summary>
+        /// Remove a rule by name.
```
Comment (baywet) [inline src/Microsoft.OpenApi/Validations/ValidationRuleSet.cs:138]:
```suggestion
        /// Remove a rule by its name from all types it is used by.
```

**Review by baywet — CHANGES_REQUESTED** [review]
Thanks for making the changes, minor comments edits I'll commit right-away

**Review by baywet — APPROVED** [review]

**Pull request merged**